modules/diagnostics/http_handler.go

```diff
@@ -0,0 +1,175 @@
+type ListControlPlaneItem struct {
+	Namespace string `json:"namespace"`
+	Name      string `json:"name"`
+	ID        string `json:"id"`
+}
+
+func (h *HTTPHandler) handleListControlPlanes(rw http.ResponseWriter, r *http.Request, cl client.Client) {
+	// List all ControlPlane CRDs.
+	cpList := operatorv2alpha1.ControlPlaneList{}
+	err := cl.List(r.Context(), &cpList)
+	if err != nil {
+		// TODO: log the error instead of directly return to the client.
+		_, _ = rw.Write([]byte(err.Error()))
+		rw.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	// List all CP instances registered to the exposer and filter CPs by their UIDs.
+	cpIDs := h.exposer.listInstances()
+	cpIDMap := lo.SliceToMap(cpIDs, func(id manager.ID) (string, struct{}) {
+		return id.String(), struct{}{}
+	})
+	// TODO: list only CPs that enabled config dump?
+	managedCPs := lo.Filter(cpList.Items, func(cp operatorv2alpha1.ControlPlane, _ int) bool {
+		_, ok := cpIDMap[string(cp.UID)]
+		return ok
+	})
+
+	// Make up the response from the filtered ControlPlanes.
+	resp := &ListControlPlanesResponse{
+		ControlPlanes: lo.Map(managedCPs, func(cp operatorv2alpha1.ControlPlane, _ int) ListControlPlaneItem {
+			return ListControlPlaneItem{
+				Namespace: cp.Namespace,
+				Name:      cp.Name,
+				ID:        string(cp.UID),
+			}
+		}),
+	}
+	rw.Header().Set("Content-Type", "application/json")
+	_ = json.NewEncoder(rw).Encode(resp)
+}
+
+func (h *HTTPHandler) handleAllControlPlanes(rw http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		rw.WriteHeader(http.StatusMethodNotAllowed)
+	}
+
+	h.handleListControlPlanes(rw, r, h.cl)
+}
+
+func (h *HTTPHandler) handleControlPlanesByNamespace(rw http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		rw.WriteHeader(http.StatusMethodNotAllowed)
+	}
+
+	namespace := r.PathValue("namespace")
+	if namespace == "" {
+		rw.WriteHeader(http.StatusBadRequest)
+		_, _ = rw.Write([]byte("empty namespace"))
+	}
+
+	clientNamespaced := client.NewNamespacedClient(h.cl, namespace)
+	h.handleListControlPlanes(rw, r, clientNamespaced)
+}
+
+func (h *HTTPHandler) handleControlPlaneConfigDump(rw http.ResponseWriter, r *http.Request) {
+	namespace := r.PathValue("namespace")
+	name := r.PathValue("name")
+
+	if namespace == "" {
+		rw.WriteHeader(http.StatusBadRequest)
+		_, _ = rw.Write([]byte("empty namespace"))
+		return
+	}
+
+	if name == "" {
+		rw.WriteHeader(http.StatusBadRequest)
+		_, _ = rw.Write([]byte("empty name"))
+		return
+	}
+
+	cp := &operatorv2alpha1.ControlPlane{}
+	err := h.cl.Get(r.Context(), client.ObjectKey{
+		Namespace: namespace,
+		Name:      name,
+	}, cp)
+
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			rw.WriteHeader(http.StatusNotFound)
+			_, _ = fmt.Fprintf(rw, "ControlPlane %s/%s not found", namespace, name)
+			return
+		}
+		// TODO: log the error instead of directly return to the client.
```

Comment:
Can we add this log or add a link to the tracking issue here?

Reply:
Added logs.